src/hardware/mixer.cpp

```diff
@@ -2767,77 +2767,83 @@ void MIXER_CloseAudioDevice()
-		// frequency changes because we've only seen them go beyond
-		// SDL's desired 48 Khz limit
-#if defined(SDL_AUDIO_ALLOW_SAMPLES_CHANGE)
-		sdl_allow_flags |= SDL_AUDIO_ALLOW_SAMPLES_CHANGE;
-#endif
+	if (allow_negotiate) {
+		// Allow negotiating the sample rate and the audio buffer size
+		// (hopefully to match the audio hardware's native rate to
+		// prevent resampling the final output, and an ideal blocksize
+		// that achieves stutter-free playback at a low latency).
+		//
+		sdl_allow_flags = SDL_AUDIO_ALLOW_SAMPLES_CHANGE |
+		                  SDL_AUDIO_ALLOW_FREQUENCY_CHANGE;
 	}
 
 	constexpr auto SdlError = 0;
 	if ((mixer.sdldevice = SDL_OpenAudioDevice(
-	             nullptr, 0, &spec, &obtained, sdl_allow_flags)) == SdlError) {
+	             nullptr, 0, &desired, &obtained, sdl_allow_flags)) == SdlError) {
+
 		LOG_WARNING("MIXER: Can't open audio device: '%s'; sound output disabled",
 		            SDL_GetError());
 
+		set_section_property_value("mixer", "nosound", "off");
 		return false;
 	}
 
-	// Does SDL want something other than stereo output?
-	if (obtained.channels != spec.channels) {
-		E_Exit("SDL gave us %d-channel output but we require %d channels",
-		       obtained.channels,
-		       spec.channels);
-	}
+	// Opening SDL audio device succeeded
+	const auto obtained_sample_rate_hz = obtained.freq;
+	const auto obtained_blocksize      = obtained.samples;
+
+	mixer.sample_rate_hz = obtained_sample_rate_hz;
+	mixer.blocksize      = obtained_blocksize;
+
+	assert(obtained.channels == NumStereoChannels);
+	assert(obtained.format == desired.format);
 
-	// Does SDL want a different playback rate?
-	if (obtained.freq != mixer.sample_rate_hz) {
-		LOG_WARNING("MIXER: SDL changed the requested sample rate of %d to %d Hz",
-		            mixer.sample_rate_hz.load(),
-		            obtained.freq);
+	// Did SDL negotiate a different playback rate?
+	if (obtained_sample_rate_hz != requested_sample_rate_hz) {
+		LOG_WARNING("MIXER: SDL negotiated the requested sample rate of %d to %d Hz",
+		            requested_sample_rate_hz,
+		            obtained_sample_rate_hz);
 
-		mixer.sample_rate_hz = obtained.freq;
 		set_section_property_value("mixer",
 		                           "rate",
 		                           format_str("%d",
 		                                      mixer.sample_rate_hz.load()));
 	}
 
-	// Does SDL want a different blocksize?
-	const auto obtained_blocksize = obtained.samples;
-
-	if (obtained_blocksize != mixer.blocksize) {
-		LOG_WARNING("MIXER: SDL changed the requested blocksize of %d to %d frames",
-		            mixer.blocksize,
-		            obtained_blocksize);
+	// Did SDL negotiate a different blocksize?
+	if (obtained_blocksize != requested_blocksize_in_frames) {
+		LOG_WARNING(
+		        "MIXER: SDL negotiated the requested blocksize of "
+		        "%d to %d frames",
+		        requested_blocksize_in_frames,
+		        obtained_blocksize);
```

Comment:
This is likely to happen on most Windows systems if negotiate is turned on (but it is turned off on Windows by default).  Up to you if you want to leave this warning in or maybe demote it to a normal log message.

Reply:
Yeah, normal message is better, cheers.

Reply:
Actually, maybe do not auto-default to negotiate the buffer size on Windows either and see what happens. I don't think setting arbitrary buffer sizes can fail either starting from Windows 10.

Reply:
Negotiate is off by default on Windows already.  It's on by default on other platforms.  It's been like that for a long time.

Reply:
Yep, when I originally added the negotiate logic, it defaulted to on across the board. Issues came up on some Windows systems, which is when kcgen added the Win-specific defaults.

Reply:
@kklobe Makes sense, but do you remember the specific use case? I could be some dude had problems on Win XP which had a very different audio API, much less lenient.

I'm inclined to "ask for forgiveness later"; turn `blocksize` negotiate off by default and see what happens. I assume noboby will run into an issue on Windows 10+.

Reply:
@johnnovak You keep saying you're going to turn negotiate off by default when that's already the case on Windows.

Reply:
@weirddan455 Indeed 😩 I find these OS specific details confusing as fuck. Okay, I'm getting fed up with all this, I'll just leave the defaults and `negotiate` alone for now 🤷🏻 It will just negotiate the block size, that's it. Maybe change it later if we need to.

Reply:
@johnnovak that was all done by kcgen; I did a quick search through the closed issues and didn't see anything specific. My original intent (https://github.com/dosbox-staging/dosbox-staging/pull/1299) was to wipe out OS-specific configs and let SDL interact with the audio driver for "ideal" sample rate and blocksize, but I suppose there are drivers out there that just didn't give back the correct settings for our callback performance requirements.

Reply:
Yeah just leave the blocksize stuff alone for now IMO.  I'm planning to take a look at the mixer to see if I can offload most of the work to a separate thread and hopefully make this a non-issue.  We can re-evaluate then.

Ideally, we would be able to accept any reasonable blocksize but let's leave it alone until we fully understand the problem.

Reply:
Sounds like a plan @weirddan455 👍🏻